src/state.rs

```diff
@@ -48,258 +41,120 @@ pub struct Constants {
 
-        self.set_minters(minters)
-    }
-
-    pub fn minters(&mut self) -> Vec<HumanAddr> {
-        self.as_readonly().minters()
-    }
-
-    pub fn tx_count(&self) -> u64 {
-        self.as_readonly().tx_count()
-    }
-
-    pub fn set_tx_count(&mut self, count: u64) -> StdResult<()> {
-        set_bin_data(&mut self.storage, KEY_TX_COUNT, &count)
+        MINTERS.save(store, &loaded_minters)
     }
 }
 
-/// This struct refactors out the readonly methods that we need for `Config` and `ReadonlyConfig`
-/// in a way that is generic over their mutability.
-///
-/// This was the only way to prevent code duplication of these methods because of the way
-/// that `ReadonlyPrefixedStorage` and `PrefixedStorage` are implemented in `cosmwasm-std`
-struct ReadonlyConfigImpl<'a, S: ReadonlyStorage>(&'a S);
-
-impl<'a, S: ReadonlyStorage> ReadonlyConfigImpl<'a, S> {
-    fn constants(&self) -> StdResult<Constants> {
-        let consts_bytes = self
-            .0
-            .get(KEY_CONSTANTS)
-            .ok_or_else(|| StdError::generic_err("no constants stored in configuration"))?;
-        bincode2::deserialize::<Constants>(&consts_bytes)
-            .map_err(|e| StdError::serialize_err(type_name::<Constants>(), e))
-    }
-
-    fn total_supply(&self) -> u128 {
-        let supply_bytes = self
-            .0
-            .get(KEY_TOTAL_SUPPLY)
-            .expect("no total supply stored in config");
-        // This unwrap is ok because we know we stored things correctly
-        slice_to_u128(&supply_bytes).unwrap()
+pub static TX_COUNT: Item<u64> = Item::new(KEY_TX_COUNT);
+pub struct TxCountStore {}
+impl TxCountStore {
+    pub fn may_load(store: &dyn Storage) -> StdResult<u64> {
+        TX_COUNT
+            .may_load(store)?
+            .ok_or_else(|| StdError::generic_err(""))
     }
 
-    fn contract_status(&self) -> ContractStatusLevel {
-        let supply_bytes = self
-            .0
-            .get(KEY_CONTRACT_STATUS)
-            .expect("no contract status stored in config");
-
-        // These unwraps are ok because we know we stored things correctly
-        let status = slice_to_u8(&supply_bytes).unwrap();
-        u8_to_status_level(status).unwrap()
-    }
-
-    fn minters(&self) -> Vec<HumanAddr> {
-        get_bin_data(self.0, KEY_MINTERS).unwrap()
-    }
-
-    pub fn tx_count(&self) -> u64 {
-        get_bin_data(self.0, KEY_TX_COUNT).unwrap_or_default()
+    pub fn save(store: &mut dyn Storage, count: u64) -> StdResult<()> {
+        TX_COUNT.save(store, &count)
     }
 }
 
-// Balances
-
-pub struct ReadonlyBalances<'a, S: ReadonlyStorage> {
-    storage: ReadonlyPrefixedStorage<'a, S>,
-}
-
-impl<'a, S: ReadonlyStorage> ReadonlyBalances<'a, S> {
-    pub fn from_storage(storage: &'a S) -> Self {
-        Self {
-            storage: ReadonlyPrefixedStorage::new(PREFIX_BALANCES, storage),
-        }
+// elad: maybe use Uint128,
+pub static BALANCES: Keymap<Addr, u128> = Keymap::new(PREFIX_BALANCES);
```

Comment:
Using `Keymap` here is interesting.
Generally, I think there's no point in using `Keymap` since the contract does not iterate over this collection at any point.
However, we did previously have cases where it could be useful, For example, creating a query to list all addresses.
Since it's pretty usecase specific, I think the best approach is for the reference implementation to not offer any iterating features (i.e. use `Item` instead).

Any thoughts on that?
cc @reuvenpo @Cashmaney

Reply:
Are you thinking of using Item with the method add_suffix() ?
It's the only way I can think of to allow us to save for each address its balance (in the case of BALANCES).

I think we have here a tradeoff of restricting functionality VS conserving space and runtime.
Because a usage with Item will keep for each address the key of: 
PREFIX_BALANCES + Addr

Reply:
The problem is redundantly keeping for each user the PREFIX_BALANCES for each data inserted.
IMO it's less correct.

Reply:
@eladr7 that's a false assumtion because when you use `Keymap` the keys are prefixed as well.

@Cashmaney do you have an idea in mind? With `Keymap` you access a key directly, so I can't see how you prevent the storage analysis.

IMO if we don't add any iteration functionality, using `Keymap` doesn't give any advantage and is just confusing. Using `Item` makes more sense I think